File: Packages/StrandAnalytics/Sources/StrandAnalytics/PersistentShiftDetector.swift
import Foundation

// PersistentShiftDetector.swift - generic one-sided longitudinal baseline-shift detector.
//
// Pure, deterministic, domain-neutral CUSUM over a robust personal baseline.
// It does not name illness, overtraining, stress, or any other cause. A caller
// chooses whether an upward or downward shift is meaningful for its metric and
// combines this temporal evidence with the domain's existing context/confounders.
public enum PersistentShiftDetector {

    public static let defaultBaselineWindow: Int = 28
    public static let defaultMinimumBaseline: Int = 7
    public static let defaultReferenceK: Double = 0.5
    public static let defaultDecisionH: Double = 4.0
    public static let defaultPersistObservations: Int = 2
    public static let defaultRecoveryZ: Double = 0.5
    public static let defaultRecoveryObservations: Int = 2
    /// Makes MAD comparable to standard deviation for a normal distribution.
    public static let normalizedMADScale: Double = 1.482602218505602

    public enum Direction: String, Equatable, Sendable {
        case upper
        case lower
    }

    public enum State: String, Equatable, Sendable {
        case missing
        case calibrating
        case degenerateBaseline
        case normal
        case watch
        case sustained
    }

    public struct Point: Equatable, Sendable {
        public let index: Int
        public let state: State
        /// Standardized deviation oriented so positive always means movement
        /// in the requested shift direction.
        public let orientedZ: Double?
        public let cusum: Double?
        public let baselineMedian: Double?
        public let baselineScale: Double?
        public let baselineCount: Int
        public let observed: Bool

        public init(index: Int, state: State, orientedZ: Double?, cusum: Double?,
                    baselineMedian: Double?, baselineScale: Double?,
                    baselineCount: Int, observed: Bool) {
            self.index = index
            self.state = state
            self.orientedZ = orientedZ
            self.cusum = cusum
            self.baselineMedian = baselineMedian
            self.baselineScale = baselineScale
            self.baselineCount = baselineCount
            self.observed = observed
        }
    }

    /// Run a one-sided Page-style CUSUM over a time-ordered optional series.
    ///
    /// `nil` is a missing observation. It emits `.missing` and neither advances
    /// nor resets the detector. Non-nil non-finite values poison the input and
    /// return nil rather than being silently treated as missing.
    ///
    /// The trailing baseline uses the median plus normalized MAD. If MAD is
    /// exactly zero, sample SD is used as a fallback. If both are zero, the
    /// point is `.degenerateBaseline` and does not accumulate a fabricated z.
    public static func evaluate(
        values: [Double?],
        direction: Direction,
        baselineWindow: Int = defaultBaselineWindow,
        minimumBaseline: Int = defaultMinimumBaseline,
        referenceK: Double = defaultReferenceK,
        decisionH: Double = defaultDecisionH,
        persistObservations: Int = defaultPersistObservations,
        recoveryZ: Double = defaultRecoveryZ,
        recoveryObservations: Int = defaultRecoveryObservations
    ) -> [Point]? {
        guard baselineWindow >= minimumBaseline,
              minimumBaseline >= 2,
              referenceK.isFinite, referenceK >= 0,
              decisionH.isFinite, decisionH > 0,
              persistObservations > 0,
              recoveryZ.isFinite,
              recoveryObservations > 0,
              values.allSatisfy({ $0 == nil || $0!.isFinite }) else { return nil }

        var out: [Point] = []
        out.reserveCapacity(values.count)
        var cusum = 0.0
        var alertRun = 0
        var recoveryRun = 0

        for i in values.indices {
            let lo = max(0, i - baselineWindow)
            var baseline: [Double] = []
            baseline.reserveCapacity(i - lo)
            if lo < i {
                for j in lo..<i {
                    if let value = values[j] { baseline.append(value) }
                }
            }

            guard let current = values[i] else {
                out.append(Point(index: i, state: .missing,
                                 orientedZ: nil, cusum: nil,
                                 baselineMedian: nil, baselineScale: nil,
                                 baselineCount: baseline.count, observed: false))
                continue
            }

            guard baseline.count >= minimumBaseline else {
                out.append(Point(index: i, state: .calibrating,
                                 orientedZ: nil, cusum: nil,
                                 baselineMedian: nil, baselineScale: nil,
                                 baselineCount: baseline.count, observed: true))
                continue
            }

            let location = median(baseline)
            let absoluteDeviations = baseline.map { abs($0 - location) }
            var scale = normalizedMADScale * median(absoluteDeviations)
            if scale <= 0 {
                scale = sampleSD(baseline)
            }
            guard scale.isFinite, scale > 0 else {
                out.append(Point(index: i, state: .degenerateBaseline,
                                 orientedZ: nil, cusum: nil,
                                 baselineMedian: location, baselineScale: nil,
                                 baselineCount: baseline.count, observed: true))
                continue
            }

            let rawZ = (current - location) / scale
            let z = direction == .upper ? rawZ : -rawZ
            cusum = max(0, cusum + z - referenceK)

            // Recovery is based on the oriented standardized deviation returning
            // near/below baseline for consecutive observed points. Missing and
            // unevaluable points never count toward recovery.
            if z < recoveryZ {
                recoveryRun += 1
                if recoveryRun >= recoveryObservations {
                    cusum = 0
                    alertRun = 0
                }
            } else {
                recoveryRun = 0
            }

            let state: State
            if cusum > decisionH {
                alertRun += 1
                state = alertRun >= persistObservations ? .sustained : .watch
            } else {
                alertRun = 0
                state = .normal
            }

            out.append(Point(index: i, state: state,
                             orientedZ: z, cusum: cusum,
                             baselineMedian: location, baselineScale: scale,
                             baselineCount: baseline.count, observed: true))
        }
        return out
    }

    private static func median(_ values: [Double]) -> Double {
        let sorted = values.sorted()
        let mid = sorted.count / 2
        if sorted.count.isMultiple(of: 2) {
            return (sorted[mid - 1] + sorted[mid]) / 2.0
        }
        return sorted[mid]
    }

    private static func sampleSD(_ values: [Double]) -> Double {
        guard values.count >= 2 else { return 0 }
        let mean = values.reduce(0, +) / Double(values.count)
        let ss = values.reduce(0) { partial, value in
            let d = value - mean
            return partial + d * d
        }
        return (ss / Double(values.count - 1)).squareRoot()
    }
}

File: android/app/src/main/java/com/noop/analytics/PersistentShiftDetector.kt
package com.noop.analytics

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.sqrt

/*
 * PersistentShiftDetector.kt - generic one-sided longitudinal baseline-shift detector.
 * Behavioral twin of StrandAnalytics/PersistentShiftDetector.swift.
 *
 * Domain-neutral CUSUM over a robust personal baseline. It names no illness,
 * stress, overtraining, or diagnosis; callers supply the domain interpretation.
 */
object PersistentShiftDetector {
    const val DEFAULT_BASELINE_WINDOW = 28
    const val DEFAULT_MINIMUM_BASELINE = 7
    const val DEFAULT_REFERENCE_K = 0.5
    const val DEFAULT_DECISION_H = 4.0
    const val DEFAULT_PERSIST_OBSERVATIONS = 2
    const val DEFAULT_RECOVERY_Z = 0.5
    const val DEFAULT_RECOVERY_OBSERVATIONS = 2
    const val NORMALIZED_MAD_SCALE = 1.482602218505602

    enum class Direction { UPPER, LOWER }
    enum class State { MISSING, CALIBRATING, DEGENERATE_BASELINE, NORMAL, WATCH, SUSTAINED }

    data class Point(
        val index: Int,
        val state: State,
        val orientedZ: Double?,
        val cusum: Double?,
        val baselineMedian: Double?,
        val baselineScale: Double?,
        val baselineCount: Int,
        val observed: Boolean,
    )

    fun evaluate(
        values: List<Double?>,
        direction: Direction,
        baselineWindow: Int = DEFAULT_BASELINE_WINDOW,
        minimumBaseline: Int = DEFAULT_MINIMUM_BASELINE,
        referenceK: Double = DEFAULT_REFERENCE_K,
        decisionH: Double = DEFAULT_DECISION_H,
        persistObservations: Int = DEFAULT_PERSIST_OBSERVATIONS,
        recoveryZ: Double = DEFAULT_RECOVERY_Z,
        recoveryObservations: Int = DEFAULT_RECOVERY_OBSERVATIONS,
    ): List<Point>? {
        if (baselineWindow < minimumBaseline ||
            minimumBaseline < 2 ||
            !referenceK.isFinite() || referenceK < 0.0 ||
            !decisionH.isFinite() || decisionH <= 0.0 ||
            persistObservations <= 0 ||
            !recoveryZ.isFinite() ||
            recoveryObservations <= 0 ||
            values.any { it != null && !it.isFinite() }
        ) return null

        val out = ArrayList<Point>(values.size)
        var cusum = 0.0
        var alertRun = 0
        var recoveryRun = 0

        for (i in values.indices) {
            val lo = max(0, i - baselineWindow)
            val baseline = ArrayList<Double>(i - lo)
            for (j in lo until i) values[j]?.let { baseline.add(it) }

            val current = values[i]
            if (current == null) {
                out.add(Point(i, State.MISSING, null, null, null, null, baseline.size, false))
                continue
            }
            if (baseline.size < minimumBaseline) {
                out.add(Point(i, State.CALIBRATING, null, null, null, null, baseline.size, true))
                continue
            }

            val location = median(baseline)
            var scale = NORMALIZED_MAD_SCALE * median(baseline.map { abs(it - location) })
            if (scale <= 0.0) scale = sampleSD(baseline)
            if (!scale.isFinite() || scale <= 0.0) {
                out.add(Point(i, State.DEGENERATE_BASELINE, null, null, location, null, baseline.size, true))
                continue
            }

            val rawZ = (current - location) / scale
            val z = if (direction == Direction.UPPER) rawZ else -rawZ
            cusum = max(0.0, cusum + z - referenceK)

            if (z < recoveryZ) {
                recoveryRun++
                if (recoveryRun >= recoveryObservations) {
                    cusum = 0.0
                    alertRun = 0
                }
            } else {
                recoveryRun = 0
            }

            val state = if (cusum > decisionH) {
                alertRun++
                if (alertRun >= persistObservations) State.SUSTAINED else State.WATCH
            } else {
                alertRun = 0
                State.NORMAL
            }

            out.add(Point(i, state, z, cusum, location, scale, baseline.size, true))
        }
        return out
    }

    private fun median(values: List<Double>): Double {
        val sorted = values.sorted()
        val mid = sorted.size / 2
        return if (sorted.size % 2 == 0) (sorted[mid - 1] + sorted[mid]) / 2.0 else sorted[mid]
    }

    private fun sampleSD(values: List<Double>): Double {
        if (values.size < 2) return 0.0
        val mean = values.sum() / values.size.toDouble()
        var ss = 0.0
        for (value in values) {
            val d = value - mean
            ss += d * d
        }
        return sqrt(ss / (values.size - 1).toDouble())
    }
}
